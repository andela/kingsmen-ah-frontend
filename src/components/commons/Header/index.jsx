import React, { Component, Fragment } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import classname from 'classnames';
import logo from '@base/img/logo.png';
import { faSearch } from '@fortawesome/fontawesome-free-solid';
import Button from '../utilities/Button';
import FontAwesome from '../utilities/FontAwesome';

export default class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      hidden: true,
      authHidden: true,
      showSearchBar: false
    };
  }

  authHeaderButtons = avatar => {
    return (
      <Fragment>
        <div className='flex items-center'>
          <FontAwesome
            type={faSearch}
            styleClass='m-2 text-gray-500 mr-4 cursor-pointer text-2xl'
            role='presentation'
            onKeyDown={this.showSearchBar}
            onClick={this.showSearchBar}
          />
        </div>

        <img
          src={avatar}
          alt='ProfileImage'
          className='rounded-full w-10 h-10 text-blue-600 cursor-pointer block'
          role='presentation'
          onKeyDown={this.showProfileDropDown}
          onClick={this.showProfileDropDown}
        />
      </Fragment>
    );
  };

  guestHeaderButtons = () => (
    <Fragment>
      <Button text='Sign In' type='regular' color='blue' onClick={() => {}} />
      <Button
        text='Get Started'
        type='outlined'
        color='blue'
        onClick={() => {}}
      />
    </Fragment>
  );

  showSearchBar = () => {
    const { showSearchBar } = this.state;

    this.setState({
      showSearchBar: !showSearchBar,
      authHidden: showSearchBar ? true : false
    });
  };

  toggleHeader = () => {
    const { hidden } = this.state;
    this.setState({
      hidden: !hidden
    });
  };

  showProfileDropDown = () => {
    const { authHidden, showSearchBar } = this.state;
    this.setState({
      authHidden: showSearchBar ? false : !authHidden,
      showSearchBar: false
    });
  };

  render() {
    const { hidden, authHidden, showSearchBar } = this.state;
    const { user, profile } = this.props;
    const { avatar, firstname, lastname } = profile;
    const { username, isAuthenticated } = user;

    return (
      <Fragment>
        <div className='bg-white shadow relative'>
          <div className='container mx-auto px-4'>
            <div className='flex items-center justify-between py-4'>
              <div className='flex'>
                <Link to='/'>
                  <img
                    src={logo}
                    alt="Authors' Haven Icon"
                    className='w-10 h-9 mr-2'
                  />
                </Link>
                <h1 className='font-bold text-xl sm-text-base'>
                  Author&lsquo;s Haven
                </h1>
              </div>

              <div className='hidden sm:flex sm:items-center'>
                {!isAuthenticated
                  ? this.guestHeaderButtons()
                  : this.authHeaderButtons(avatar)}
              </div>

              <div className='flex sm:hidden cursor-pointer'>
                {!isAuthenticated ? (
                  <svg
                    xmlns='http://www.w3.org/2000/svg'
                    viewBox='0 0 512 512'
                    className='w-6 h-6 text-blue-600 cursor-pointer'
                    onClick={this.toggleHeader}
                  >
                    <path d='M64 384h384v-42.666H64V384zm0-106.666h384v-42.667H64v42.667zM64 128v42.665h384V128H64z' />
                  </svg>
                ) : (
                  this.authHeaderButtons(avatar)
                )}
              </div>
            </div>

            <div
              className={classname({
                'block sm:hidden bg-white border-t-2 py-2': true,
                hidden
              })}
            >
              <div className='flex justify-between items-center pt-2'>
                {!isAuthenticated ? this.guestHeaderButtons(avatar) : ''}
              </div>
            </div>
          </div>
        </div>

        <div
          className={classname({
            'block bg-white border-t-2 py-2 w-2/5 max-w-xs shadow-lg absolute right-0': true,
            hidden: authHidden
          })}
        >
          <div className='tooltip container' />
          <div className='flex justify-between items-center pt-2 font-sans text-sm'>
            {isAuthenticated && !showSearchBar ? (
              <div className='p-2 px-2 sm:px-6  w-full'>
                <Link to={`/profile/${username}`}>
                  <div className='border-b pb-4 md:flex lg:flex flex-wrap'>
                    <img
                      src={avatar}
                      alt='My profile'
                      className='rounded-full w-10 h-10 mx-auto md:mx-0 lg:mx-0'
                    />
                    <div className='ml-0 md:ml-4 lg:ml-4 md:text-left lg:text-left text-center hover:text-blue-700'>
                      <div className='font-bold text-base'>{`${firstname} ${lastname}`}</div>
                      <div className='text-gray-500'>{`@ ${username}`}</div>
                    </div>
                  </div>
                </Link>

                <Link
                  to={`/profile/${username}`}
                  className='block border-b pb-2 hover:text-blue-700 pt-2'
                >
                  Profile
                </Link>
                <Link
                  to='/articles'
                  className='block border-b pb-2 pt-2 hover:text-blue-700'
                >
                  New Article
                </Link>
                <Link to='/' className='block pt-2 hover:text-blue-700'>
                  Sign out
                </Link>
              </div>
            ) : (
              ''
            )}

            {isAuthenticated && showSearchBar ? (
              <input
                type='text'
                className='mr-4 resize-x w-full ml-4 mb-2'
                placeholder='Search...'
                autoFocus
              />
            ) : (
              ''
            )}
          </div>
        </div>
      </Fragment>
    );
  }
}

Header.propTypes = {
  user: PropTypes.shape({
    isAuthenticated: PropTypes.bool,
    username: PropTypes.string
  }).isRequired,
  profile: PropTypes.shape({
    firstname: PropTypes.string,
    lastname: PropTypes.string,
    avatar: PropTypes.string
  }).isRequired
};
