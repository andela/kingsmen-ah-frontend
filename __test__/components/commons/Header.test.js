import React from 'react';
import Header from '@components/commons/Header/index';
import { logo } from '@base/img/logo.png';

const props = {
  user: {
    isAuthenticated: true,
    username: 'kingsmen'
  },
  profile: {
    firstname: 'Kings',
    lastname: 'men',
    avatar: logo
  }
};

const setup = () => {
  const wrapper = shallow(<Header {...props} />);
  return wrapper;
};

describe('<Header />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = setup();
  });

  it('should render the Header component correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('Header')).toBeTruthy();
  });
});
