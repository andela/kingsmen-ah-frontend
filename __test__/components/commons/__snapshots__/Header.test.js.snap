// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render the Header component correctly 1`] = `
<Fragment>
  <div
    className="bg-white shadow relative"
  >
    <div
      className="container mx-auto px-4"
    >
      <div
        className="flex items-center justify-between py-4"
      >
        <div
          className="flex"
        >
          <Link
            to="/"
          >
            <img
              alt="Authors' Haven Icon"
              className="w-10 h-9 mr-2"
              src={
                Object {
                  "process": [Function],
                }
              }
            />
          </Link>
          <h1
            className="font-bold text-xl sm-text-base"
          >
            Authorâ€˜s Haven
          </h1>
        </div>
        <div
          className="hidden sm:flex sm:items-center"
        >
          <div
            className="flex items-center"
          >
            <FontAwesome
              onClick={[Function]}
              onKeyDown={[Function]}
              role="presentation"
              styleClass="m-2 text-gray-500 mr-4 cursor-pointer text-2xl"
              type={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f002",
                    "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
                  ],
                  "iconName": "search",
                  "prefix": "fas",
                }
              }
            />
          </div>
          <img
            alt="ProfileImage"
            className="rounded-full w-10 h-10 text-blue-600 cursor-pointer block"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="presentation"
          />
        </div>
        <div
          className="flex sm:hidden cursor-pointer"
        >
          <div
            className="flex items-center"
          >
            <FontAwesome
              onClick={[Function]}
              onKeyDown={[Function]}
              role="presentation"
              styleClass="m-2 text-gray-500 mr-4 cursor-pointer text-2xl"
              type={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f002",
                    "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
                  ],
                  "iconName": "search",
                  "prefix": "fas",
                }
              }
            />
          </div>
          <img
            alt="ProfileImage"
            className="rounded-full w-10 h-10 text-blue-600 cursor-pointer block"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="presentation"
          />
        </div>
      </div>
      <div
        className="block sm:hidden bg-white border-t-2 py-2 hidden"
      >
        <div
          className="flex justify-between items-center pt-2"
        />
      </div>
    </div>
  </div>
  <div
    className="block bg-white border-t-2 py-2 w-2/5 max-w-xs shadow-lg absolute right-0 hidden"
  >
    <div
      className="tooltip container"
    />
    <div
      className="flex justify-between items-center pt-2 font-sans text-sm"
    >
      <div
        className="p-2 px-2 sm:px-6  w-full"
      >
        <Link
          to="/profile/kingsmen"
        >
          <div
            className="border-b pb-4 md:flex lg:flex flex-wrap"
          >
            <img
              alt="My profile"
              className="rounded-full w-10 h-10 mx-auto md:mx-0 lg:mx-0"
            />
            <div
              className="ml-0 md:ml-4 lg:ml-4 md:text-left lg:text-left text-center hover:text-blue-700"
            >
              <div
                className="font-bold text-base"
              >
                Kings men
              </div>
              <div
                className="text-gray-500"
              >
                @ kingsmen
              </div>
            </div>
          </div>
        </Link>
        <Link
          className="block border-b pb-2 hover:text-blue-700 pt-2"
          to="/profile/kingsmen"
        >
          Profile
        </Link>
        <Link
          className="block border-b pb-2 pt-2 hover:text-blue-700"
          to="/articles"
        >
          New Article
        </Link>
        <Link
          className="block pt-2 hover:text-blue-700"
          to="/"
        >
          Sign out
        </Link>
      </div>
    </div>
  </div>
</Fragment>
`;
